app/src/main/java/uk/gov/android/authentication/AppAuthSession.kt

```diff
@@ -0,0 +1,117 @@
+package uk.gov.android.authentication
+
+import android.app.Activity
+import android.content.Context
+import android.content.Intent
+import androidx.core.app.ActivityCompat
+import java.util.UUID
+import net.openid.appauth.AuthorizationException
+import net.openid.appauth.AuthorizationRequest
+import net.openid.appauth.AuthorizationResponse
+import net.openid.appauth.AuthorizationService
+import net.openid.appauth.AuthorizationServiceConfiguration
+
+@Suppress("TooGenericExceptionThrown")
+class AppAuthSession : LoginSession {
+    private var context: Context? = null
+    private lateinit var authService: AuthorizationService
+
+    override fun init(
+        context: Context
+    ): LoginSession {
+        if (this.context == null) {
+            this.context = context
+            authService = AuthorizationService(context)
+        }
+        return this
+    }
+
+    override fun present(
+        configuration: LoginSessionConfiguration
+    ) {
+        if (context == null) {
+            throw Error("Context is null, did you call init?")
+        }
+
+        with(configuration) {
+            val context = this@AppAuthSession.context!!
+            val nonce = UUID.randomUUID().toString()
+
+            val serviceConfig = AuthorizationServiceConfiguration(
+                authorizeEndpoint,
+                tokenEndpoint
+            )
+
+            val builder = AuthorizationRequest.Builder(
+                serviceConfig,
+                clientId,
+                responseType,
+                redirectUri
+            ).also {
+                it.apply {
+                    setScopes(scopes)
+                    setUiLocales(locale)
+                    setNonce(nonce)
+                    setAdditionalParameters(
+                        mapOf(
+                            "vtr" to vectorsOfTrust
+                        )
+                    )
+                }
+            }
+
+            val authRequest = builder.build()
+
+            val authIntent = authService.getAuthorizationRequestIntent(authRequest)
+            ActivityCompat.startActivityForResult(
+                context as Activity,
+                authIntent,
+                REQUEST_CODE_AUTH,
+                null
+            )
+        }
+    }
+
+    override fun finalise(intent: Intent, callback: (tokens: TokenResponse) -> Unit) {
+        val authorizationResponse = AuthorizationResponse.fromIntent(intent)
+
+        if (authorizationResponse == null) {
+            val exception = AuthorizationException.fromIntent(intent)
+
+            throw Exception(exception?.message)
```

Comment:
Will you be displaying different errors from the exceptions thrown at line 81 and 90? If so you may want to subclass the generic exception